base/shared/src/main/scala/scalaz/data/ilist.scala

```diff
@@ -37,6 +41,87 @@ trait IListModule {
   implicit def isCovariantInstance: IsCovariant[IList]
 }
 
+trait IListSyntax {
+  final class ToIListOps[A](self: IList[A]) {
+
+    def headOption: Maybe[A] =
+      IList.uncons(self) match {
+        case Maybe2.Empty2()    => Maybe.empty
+        case Maybe2.Just2(a, _) => Maybe.just(a)
+      }
+
+    def tailOption: Maybe[IList[A]] =
+      IList.uncons(self) match {
+        case Maybe2.Empty2()     => Maybe.empty
+        case Maybe2.Just2(_, as) => Maybe.just(as)
+      }
+
+    def ::(a: A): IList[A] = IList.cons(a, self)
+
+    def uncons: Maybe2[A, IList[A]] =
+      IList.uncons(self)
+
+    def foldLeft[B](z: B)(f: (B, A) => B) =
+      IList.foldLeft(f, z, self)
+
+    def append(that: IList[A]): IList[A] =
+      IList.append(self, that)
+
+    def :::[Z](that: IList[A]): IList[A] =
+      IList.foldLeft[A, IList[A]]((b, a) => IList.cons(a, b), self, IList.reverse(that))
+
+    def reverse: IList[A] =
+      IList.reverse(self)
+
+    def reverse_:::[Z](that: IList[A]): IList[A] =
+      IList.foldLeft[A, IList[A]]((as, a) => IList.cons(a, as), self, that)
+
+    def filter(p: A => Boolean): IList[A] =
+      IList.foldLeft[A, IList[A]]((b, a) => if (p(a)) IList.cons(a, b) else b, IList.empty, self)
+
+    def zip[B](that: IList[B]): IList[(A, B)] = {
+      @tailrec
+      def go(acc: IList[(A, B)], as: IList[A], bs: IList[B]): IList[(A, B)] =
+        (IList.uncons(as), IList.uncons(bs)) match {
+          case (Maybe2.Just2(a, aas), Maybe2.Just2(b, bbs)) => go(IList.cons((a, b), acc), aas, bbs)
+          case _                                            => acc
+        }
+
+      go(IList.empty, self, that)
+    }
+
+    def take(n: Int): IList[A] = {
+      @tailrec
+      def go(m: Int, acc: IList[A], as: IList[A]): IList[A] =
+        IList.uncons(as) match {
+          case Maybe2.Empty2() => IList.reverse(acc)
+          case Maybe2.Just2(a, aas) =>
+            go(m - 1, IList.cons(a, acc), aas)
+        }
+
+      if (n <= 0) IList.empty
+      else go(n, IList.empty, self)
+    }
+
+    def drop(n: Int): IList[A] = {
+      @tailrec
+      def go(m: Int, as: IList[A]): IList[A] =
+        IList.uncons(as) match {
+          case Maybe2.Empty2() => as
+          case Maybe2.Just2(_, aas) =>
+            go(m - 1, aas)
+        }
+
+      if (n <= 0) self
+      else go(n, self)
+    }
+
+    def size: Int =
+      IList.foldLeft[A, Int]((b, a) => 1 + b, 0, self)
```

Comment:
This will be faster specialized.

Reply:
Good call!

Reply:
He means that writing it out in terms of pattern matching + `@tailrec` function will be faster, because `Int`s won't be boxed and the function will be "inlined" (manually).

Reply:
oh, sure i can do that. I thought he was referring to `@specialized` and I was like 😖